r2/types.go

```diff
@@ -0,0 +1,308 @@
+	for k, v := range f {
+		tmp = append(tmp, fmt.Sprintf("%s:%s", k, v))
+	}
+	marshalled := strconv.Quote(strings.Join(tmp, " "))
+	return []byte(marshalled), nil
+}
+
+func (f *filters) UnmarshalJSON(data []byte) error {
+	str := string(data)
+	unquoted, err := strconv.Unquote(str)
+	if err != nil {
+		return err
+	}
+
+	*f = make(map[string]string)
+	tagPairs := strings.Split(unquoted, tagFilterListSeparator)
+	for _, p := range tagPairs {
+		parts := strings.Split(p, tagFilterSeparator)
+		if len(parts) != 2 {
+			return NewBadInputError(fmt.Sprintf("invalid tag filter pair %s", p))
+		}
+		(*f)[parts[0]] = parts[1]
+	}
+	return nil
+}
+
+type mappingRuleJSON struct {
+	ID             string          `json:"id,omitempty"`
+	Name           string          `json:"name" validate:"required"`
+	CutoverMillies int64           `json:"cutoverMillies,omitempty"`
+	Filters        filters         `json:"filter" validate:"required"`
+	Policies       []policy.Policy `json:"policies" validate:"required"`
+}
+
+func (m mappingRuleJSON) mappingRuleView() rules.MappingRuleView {
+	return rules.MappingRuleView{
+		ID:       m.ID,
+		Name:     m.Name,
+		Filters:  m.Filters,
+		Policies: m.Policies,
+	}
+}
+
+func newMappingRuleJSON(mrv *rules.MappingRuleView) mappingRuleJSON {
+	return mappingRuleJSON{
+		ID:             mrv.ID,
+		Name:           mrv.Name,
+		Filters:        mrv.Filters,
+		Policies:       mrv.Policies,
+		CutoverMillies: nanoToMilli(mrv.CutoverNanos),
+	}
+}
+
+type mappingRuleHistory struct { // nolint
+	States []mappingRuleJSON `json:"states"`
+}
+
+// RollupTargetData ...
+type rollupTargetJSON struct {
+	Name     string          `json:"name" validate:"required"`
+	Tags     []string        `json:"tags" validate:"required"`
+	Policies []policy.Policy `json:"policies" validate:"required"`
+}
+
+func (t rollupTargetJSON) rollupTargetView() rules.RollupTargetView {
+	return rules.RollupTargetView{
+		Name:     t.Name,
+		Tags:     t.Tags,
+		Policies: t.Policies,
+	}
+}
+
+func newRollupTargetJSON(t rules.RollupTargetView) rollupTargetJSON {
+	return rollupTargetJSON{
+		Name:     t.Name,
+		Tags:     t.Tags,
+		Policies: t.Policies,
+	}
+}
+
+type rollupRuleJSON struct {
+	ID             string             `json:"id,omitempty"`
+	Name           string             `json:"name" validate:"required"`
+	Filters        filters            `json:"filter" validate:"required"`
+	Targets        []rollupTargetJSON `json:"targets" validate:"required,dive,required"`
+	CutoverMillies int64              `json:"cutoverMillies,omitempty"`
+}
+
+func newRollupRuleJSON(rrv *rules.RollupRuleView) rollupRuleJSON {
+	targets := make([]rollupTargetJSON, len(rrv.Targets))
+	for i, t := range rrv.Targets {
+		targets[i] = newRollupTargetJSON(t)
+	}
+	return rollupRuleJSON{
+		ID:             rrv.ID,
+		Name:           rrv.Name,
+		Filters:        rrv.Filters,
+		Targets:        targets,
+		CutoverMillies: nanoToMilli(rrv.CutoverNanos),
```

Comment:
why `Millies`?

Reply:
Browsers don't have  64 bit ints. Javascript would get confused if I returned a timestamp in nanos.

Reply:
If you're asking why I added the e... Read that millies was more correct somewhere and went with it.

Reply:
Yeah I was commenting on the spelling. It's better to stick to popular convention, even if grammatically incorrect for APIs. Change to `Millis`.